### PR TITLE
feat: chain spec check

### PR DESCRIPTION
**Pull request by zhangsoledad**

### CODE REVIEW

File: ckb-bin/src/subcommand/replay.rs [hunk 1]
```diff
@@ -44,7 +44,7 @@ pub fn replay(args: ReplayArgs) -> Result<(), ExitCode> {
         if let Some((from, to)) = args.profile {
             profile(shared, chain, from, to);
         } else if args.sanity_check {
-            sanity_check(shared, chain, args.full_verfication);
+            sanity_check(shared, chain, args.full_verification);
```
Comment (doitian) [inline ckb-bin/src/subcommand/replay.rs:47]: 🙄

File: util/app-config/src/cli.rs [hunk 2]
```diff
@@ -87,11 +88,18 @@ pub fn get_matches(version: &Version) -> ArgMatches<'static> {
...
+        .about("Runs ckb node")
+        .arg(
+            Arg::with_name(ARG_BA_ADVANCED)
+                .long(ARG_BA_ADVANCED)
+                .help("Allows any block assembler code hash and args"),
+        )
+        .arg(
+            Arg::with_name(ARG_SKIP_CHAIN_SPEC_CHECK)
+                .long(ARG_SKIP_CHAIN_SPEC_CHECK)
+                .help("Enforce consistency check on chain-spec and storage"),
```
Comment (doitian) [inline util/app-config/src/cli.rs:101]:
```suggestion
                .help("Skips checking the chain spec with the hash stored in the database"),
```

File: ckb-bin/src/subcommand/run.rs [hunk 3]
```diff
@@ -180,6 +182,35 @@ fn verify_genesis(shared: &Shared) -> Result<(), ExitCode> {
         })
 }
 
+fn check_spec(shared: &Shared, args: &RunArgs) -> Result<(), ExitCode> {
+    let store = shared.store();
+    if let Some(spec_hash) = store.get_chain_spec_hash() {
+        if args.chain_spec_hash != spec_hash && !args.skip_chian_spec_check {
+            eprintln!(
+                "chain_spec_hash mismatch Config({}) storage({})",
```
Comment (doitian) [inline ckb-bin/src/subcommand/run.rs:190]:
```suggestion
                "chain_spec_hash mismatch Config({}) storage({}), pass command line argument --skip-spec-check if you are sure that the two different chains are compatible.",
```

The suggested change may require `cargo fmt`

**Review by doitian — APPROVED** [review]

**Review by doitian — CHANGES_REQUESTED** [review]
Request the error message change so user know how to skip the check.

**Review by doitian — DISMISSED** [review]

**Comment by doitian** [comment]
@zhangsoledad CI failed

**Review by keroro520 — DISMISSED** [review]

**Review by keroro520 — APPROVED** [review]

**Review by doitian — APPROVED** [review]

**Comment by doitian** [comment]
bors r=doitian,keroro520

**Pull request merged**